fedimint-server-core/src/init.rs

```diff
@@ -325,9 +331,25 @@ where
             <Self as ServerModuleInit>::get_client_config(self, config)?,
         )
     }
-
-    fn get_database_migrations(&self) -> BTreeMap<DatabaseVersion, CoreMigrationFn> {
+    fn get_database_migrations(&self) -> BTreeMap<DatabaseVersion, DynServerDbMigrationFn> {
         <Self as ServerModuleInit>::get_database_migrations(self)
+            .into_iter()
+            .map(|(k, f)| {
+                (k, {
+                    let closure: DynServerDbMigrationFn =
+                        Box::new(move |ctx: ServerDbMigrationFnContext<'_>| {
+                            let map = ctx.map(ServerModuleDbMigrationContext::new);
+                            Box::pin(f(map))
+                        });
+                    // safety: Seems like it's impossible to express
+                    // this code without `for<'tx>` on closure which is an unstable
+                    // feature: https://github.com/rust-lang/rust/issues/97362
+                    unsafe {
```

Comment:
:shrug: @elsirion 

I don't really see any good way to avoid it. AFAICT the code is sound, it's just the limitation described in the linked issue.

Reply:
If I understand correctly, this is mapping `ServerDbMigrationFnContext` into a `DynServerDbMigrationFn`? And this is necessary because previously we didn't have a type-erased `DynServerDbMigrationFn` before?

Reply:
Yes, the `ServerDbMigrationFnContext` is a "context", which knows the types of history items the module wants. So we need to type erase it here to a single `DynServerDbMigrationFn` that is the same for all modules. Just like we do for a lot of methods in this

```
#[apply(async_trait_maybe_send!)]
impl<T> IServerModuleInit for T
where
    T: ServerModuleInit + 'static + Sync,
{
```

block.

The `transmute` is used because the `DbMigrationFn` expects `for<'tx> ....`, while this new closure get a signature without `for<'tx>` and there's currently no way to express it on stable Rust.